src/wp-includes/class-wp-theme-json-resolver.php

```diff
@@ -576,6 +612,26 @@ public static function get_merged_data( $origin = 'custom' ) {
 			_deprecated_argument( __FUNCTION__, '5.9.0' );
 		}
 
+		if ( ! in_array( $origin, WP_Theme_JSON::VALID_ORIGINS, true ) ) {
+			$origin = 'custom';
+		}
+
+		// Map origins to block cache values.
+		$cache_map = array(
+			'default' => 'core',
+			'blocks'  => 'blocks',
+			'theme'   => 'theme',
+			'custom'  => 'user',
+		);
+
+		if (
+			null !== static::$merged[ $origin ]
+			&& static::has_same_registered_blocks( $cache_map[ $origin ] )
+			&& static::get_theme_supports_data() === static::$theme_support_data
```

Comment:
Theme support data only matters for the `theme` and `custom` origins, so I think we should only apply this check if the given origin is one of those, to keep caches warm for the other two origins.